metadata/src/main/java/org/apache/kafka/controller/BootstrapMetadata.java

```diff
@@ -0,0 +1,198 @@
+            '}';
+    }
+
+    /**
+     * A raft client listener that simply collects all of the commits and snapshots into a mapping of
+     * metadata record type to list of records.
+     */
+    private static class BootstrapListener implements RaftClient.Listener<ApiMessageAndVersion> {
+        private final List<ApiMessageAndVersion> records = new ArrayList<>();
+
+        @Override
+        public void handleCommit(BatchReader<ApiMessageAndVersion> reader) {
+            try {
+                while (reader.hasNext()) {
+                    Batch<ApiMessageAndVersion> batch = reader.next();
+                    records.addAll(batch.records());
+                }
+            } finally {
+                reader.close();
+            }
+        }
+
+        @Override
+        public void handleSnapshot(SnapshotReader<ApiMessageAndVersion> reader) {
+            try {
+                while (reader.hasNext()) {
+                    Batch<ApiMessageAndVersion> batch = reader.next();
+                    for (ApiMessageAndVersion messageAndVersion : batch) {
+                        records.add(messageAndVersion);
+                    }
+                }
+            } finally {
+                reader.close();
+            }
+        }
+    }
+
+    public static BootstrapMetadata create(MetadataVersion metadataVersion) {
+        return create(metadataVersion, new ArrayList<>());
+    }
+
+    public static BootstrapMetadata create(MetadataVersion metadataVersion, List<ApiMessageAndVersion> records) {
+        if (!metadataVersion.isKRaftSupported()) {
+            throw new IllegalArgumentException("Cannot create BootstrapMetadata with a non-KRaft metadata version.");
+        }
+        records.add(new ApiMessageAndVersion(
+            new FeatureLevelRecord()
+                .setName(MetadataVersion.FEATURE_NAME)
+                .setFeatureLevel(metadataVersion.featureLevel()),
+            FeatureLevelRecord.LOWEST_SUPPORTED_VERSION));
+
+        return new BootstrapMetadata(metadataVersion, records);
+    }
+
+    /**
+     * Load a bootstrap snapshot into a read-only bootstrap metadata object and return it.
+     *
+     * @param bootstrapDir  The directory from which to read the snapshot file.
+     * @return              The read-only bootstrap metadata
+     * @throws Exception
+     */
+    public static BootstrapMetadata load(Path bootstrapDir) throws Exception {
+        final Path bootstrapPath = bootstrapDir.resolve(BOOTSTRAP_FILE);
+
+        if (!Files.exists(bootstrapPath)) {
+            log.debug("Missing bootstrap file, this appears to be a KRaft preview cluster.");
+            return BootstrapMetadata.create(MetadataVersion.IBP_3_0_IV0);
+        }
+
+        BootstrapListener listener = new BootstrapListener();
+        SnapshotFileReader reader = new SnapshotFileReader(bootstrapPath.toString(), listener);
+        reader.startup();
+        try {
+            reader.caughtUpFuture().get();
```

Comment:
It looks like this implementation starts a thread and blocks on that thread reading the bootstrap snapshot? Why it the thread needed? Why not read the snapshot on this thread?

Reply:
Yeah. The thread was useful for the metadata shell since the idea was it would keep handling new updates in the background. Obviously, that isn't needed here, but if we're reusing the code, we'll get it.

Long-term, we could have a version of EventQueue that uses the current thread (kind of like Java's executor that does the same thing, can't remember what it's called) and use it here.